### PR TITLE
deprecate spones

### LINKED ISSUE #22381: spones documentation is incorrect
The documentation for the spones function describes the Matlab behavior for this function, which replaces each non-zero entry by a value of 1.0. However, the implementation appears to use julia's notion of "stored entries" instead. 

~~~~
julia> versioninfo()
Julia Version 0.6.0-rc3.0
Commit ad290e9 (2017-06-07 11:53 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)

help?> spones
search: spones unsafe_pointer_to_objref

  spones(S)

  Create a sparse array with the same structure as that of S, but with every nonzero element having the value 1.0.

  julia> A = sparse([1,2,3,4],[2,4,3,1],[5.,4.,3.,2.])
  4×4 SparseMatrixCSC{Float64,Int64} with 4 stored entries:
    [4, 1]  =  2.0
    [1, 2]  =  5.0
    [3, 3]  =  3.0
    [2, 4]  =  4.0
  
  julia> spones(A)
  4×4 SparseMatrixCSC{Float64,Int64} with 4 stored entries:
    [4, 1]  =  1.0
    [1, 2]  =  1.0
    [3, 3]  =  1.0
    [2, 4]  =  1.0

  Note the difference from speye.

julia> A = sparse([1,2],[1,2],[true,false])
2×2 SparseMatrixCSC{Bool,Int64} with 2 stored entries:
  [1, 1]  =  true
  [2, 2]  =  false

julia> B = spones(A)
2×2 SparseMatrixCSC{Bool,Int64} with 2 stored entries:
  [1, 1]  =  true
  [2, 2]  =  true

julia> A = sparse([1,2],[1,2],[1.0,0.0])
2×2 SparseMatrixCSC{Float64,Int64} with 2 stored entries:
  [1, 1]  =  1.0
  [2, 2]  =  0.0

julia> C = spones(A)
2×2 SparseMatrixCSC{Float64,Int64} with 2 stored entries:
  [1, 1]  =  1.0
  [2, 2]  =  1.0
~~~~

Expected behavior given the documentation

~~~~
julia> B = spones(A)
2×2 SparseMatrixCSC{Float64,Int64} with 1 stored entries:
  [1, 1]  =  1.0

julia> C = spones(A)
2×2 SparseMatrixCSC{Float64,Int64} with 1 stored entries:
  [1, 1]  =  1.0
~~~~

So note that there are two issues with the current documentation. First, the value is one(T) not 1.0. Second, the operation proceeds over the stored entries, not the non-zero stored entries.

Ideal handling.

* Change the operation of `spones` to do the same thing as Matlab's function. In which case, this would operate according to the documentation. (That is, it would satisfy the relationship `countnz(spones(A)) == countnz(A)`)
* Introduce a new function for the current behavior. `spsones` ? 

Minimal handling.

* Adjust the documentation for the current function. Would suggest including an example describing the difference from Matlab's behavior (as above). 

I don't like the minimal handling because the semantics of Matlab's `spones` and Julia's `spones` are radically different in unexpected and bug-inducing fashions because I have no idea which Julia operations introduce new structural non-zeros. (Let's just say I spent a while working with the wrong distribution of random matrices.)

### PR DESCRIPTION
**Pull request by fredrikekre**
This is a rather useless function.

This is on top of #25030, relevant commit for this PR is a0ada0f73a69d272b996f50dafca55ef0f682a69
fix #22381

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -704,6 +704,9 @@ Deprecated or removed
     have been deprecated in favor of `spdiagm(d => x)` and `spdiagm(d[1] => x[1], d[2] => x[2], ...)`
     respectively. The new `spdiagm` implementation now always returns a square matrix ([#23757]).
 
+  * `spones(A::AbstractSparseArray{T})` has been deprecated in favor of
+    `LinAlg.fillstored!(copy(A), one(T))` ([#TBD]).
```
Comment (Sacha0) [inline NEWS.md]: The `one(T)` could simplify to `1` in the vast majority of cases, as `fillstored!` necessarily converts that `1` to `eltype(A)`? Suggesting the simplest replacements first is nice :).

File: base/deprecated.jl [hunk 2]
```diff
@@ -1861,6 +1861,12 @@ end
 # PR #25030
 @eval LinAlg @deprecate fillslots! fillstored! false
 
+# PR #TBD
+@eval SparseArrays @deprecate spones(A::SparseMatrixCSC{T}) where {T} fillstored!(copy(A), one(T))
+@eval SparseArrays @deprecate spones(A::SparseVector{T}) where {T} fillstored!(copy(A), one(T))
```
Comment (Sacha0) [inline base/deprecated.jl]: Likewise here, perhaps simplify the suggested replacement as much as possible?

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -1431,31 +1431,7 @@ julia> sprandn(rng, 2, 2, 0.75)
...
-4×4 SparseMatrixCSC{Float64,Int64} with 4 stored entries:
-  [4, 1]  =  1.0
-  [1, 2]  =  1.0
-  [3, 3]  =  1.0
-  [2, 4]  =  1.0
-```
-"""
-spones(S::SparseMatrixCSC{T}) where {T} =
-     SparseMatrixCSC(S.m, S.n, copy(S.colptr), copy(S.rowval), ones(T, S.colptr[end]-1))
+LinAlg.fillstored!(S::SparseMatrixCSC, x) = (fill!(S.nzval, x); S)
```
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: `fill!(view(S.nzval, 1:(S.colptr[S.n + 1] - 1)), x)`?

File: base/sparse/sparse.jl [hunk 4]
```diff
@@ -31,7 +31,7 @@ import Base: @get!, acos, acosd, acot, acotd, acsch, asech, asin, asind, asinh,
 
 export AbstractSparseArray, AbstractSparseMatrix, AbstractSparseVector,
     SparseMatrixCSC, SparseVector, blkdiag, droptol!, dropzeros!, dropzeros,
-    issparse, nonzeros, nzrange, rowvals, sparse, sparsevec, spdiagm, spones,
+    issparse, nonzeros, nzrange, rowvals, sparse, sparsevec, spdiagm, fillstored!,
```
Comment (Sacha0) [inline base/sparse/sparse.jl]: Why export `fillstored!` from `SparseArrays` while `LinAlg` owns `fillstored!`?

File: base/sparse/sparsevector.jl [hunk 5]
```diff
@@ -99,8 +99,7 @@ spzeros(len::Integer) = spzeros(Float64, len)
 spzeros(::Type{T}, len::Integer) where {T} = SparseVector(len, Int[], T[])
 spzeros(::Type{Tv}, ::Type{Ti}, len::Integer) where {Tv,Ti<:Integer} = SparseVector(len, Ti[], Tv[])
 
-# Construction of same structure, but with all ones
-spones(x::SparseVector{T}) where {T} = SparseVector(x.n, copy(x.nzind), ones(T, length(x.nzval)))
+LinAlg.fillstored!(x::SparseVector{T}, y) where {T} = (fill!(x.nzval, y); x)
```
Comment (Sacha0) [inline base/sparse/sparsevector.jl]: Perhaps nix the unnecessary type parameter?

File: test/sparse/sparse.jl [hunk 6]
```diff
@@ -1737,8 +1737,8 @@ end
     end
 end
 
-@testset "spones" begin
-    @test spones(sparse(2.0I, 5, 5)) == Matrix(I, 5, 5)
+@testset "fillstored!" begin
+    @test SparseArrays.fillstored!(sparse(2.0I, 5, 5), 1) == Matrix(I, 5, 5)
```
Comment (Sacha0) [inline test/sparse/sparse.jl]: Why the `SparseArrays` qualification when the definitions in `SparseArrays` extend the name from `LinAlg`?

**Comment by Sacha0** [comment]
Much thanks for clearing this one @fredrikekre! :)

**Comment by fredrikekre** [comment]
Rebased, planning to merge this when ci is done.

File: base/deprecated.jl [hunk 7]
```diff
@@ -1898,6 +1898,12 @@ end
 # PR #25030
 @eval LinAlg @deprecate fillslots! fillstored! false
 
+# PR #TBD
```
Comment (rfourquet) [inline base/deprecated.jl]: Todo! the CI jobs to seem to have started yet, so it's still time to update this without having CI resources wasted.

**Comment by StefanKarpinski** [comment]
I may have broken the build right when this was pushed, so probably worth fixing and rebasing.

File: base/deprecated.jl (hunk 7 above)
Comment (fredrikekre) [reply base/deprecated.jl]: Thanks, fixed!

**Comment by StefanKarpinski** [comment]
Nevermind, I fixed it and pushed it for you.

**Comment by StefanKarpinski** [comment]
I think we did the same thing at the same time...

**Comment by fredrikekre** [comment]
I'm so confused... I only fixed the missing PR number that Rafael pointed out :)

File: base/deprecated.jl (hunk 7 above)
Comment (rfourquet) [reply base/deprecated.jl]: In the future maybe CI will check for `#TBD` `#TODO`, `[#CATS]` etc!

**Comment by StefanKarpinski** [comment]
Ok, I did that plus I rebased on top of a newer master where I fixed the merge mistake.

**Comment by Sacha0** [comment]
The i686 AV linalg/bunchkaufman failure is identical to that in #25083. Seems something broke AV i686 linalg/bunchkaufman recently but as nonetheless merged? Having a look at the AV queue...

**Comment by Sacha0** [comment]
The AV i686 failure is #25096, fixed by #25103. Best!

**Review by Sacha0 — APPROVED** [review]
lgtm for the most recent version! Thanks @fredrikekre! :)

**Pull request merged**